### ISSUE TITLE
Postgres compatibility for date/time parsing code

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#3138]
The date/time parser was written in an era before we'd decided that Postgres compatibility was an important goal. As a result it only attempts to support relatively sane date/time formats. The Postgres parser, by contrast, supports approximately every date format ever used by humanity, in any language, in any era.

We've been adding one-off support for new date-time formats as necessary, but at some point we might want to sit down with our implementation and the Postgres implementation side by side, and see if we can make them match.

The underlying strategy of the Postgres implementation is relatively similar to our own, I think. There is a [`ParseDateTime` function](https://github.com/postgres/postgres/blob/a995b371ae29de2d38c4b7881cf414b1560e9746/src/backend/utils/adt/datetime.c#L505-L730) that is the approximate equivalent of our [`tokenize_time_str`](https://github.com/MaterializeInc/materialize/blob/9f7c5654eb05201c0aac85a31406a4837f45c540/src/repr/scalar/datetime.rs#L1171). Then there are various functions that munge that parsed representation into an actual typed object, like [`DecodeDateTime`](https://github.com/postgres/postgres/blob/a995b371ae29de2d38c4b7881cf414b1560e9746/src/backend/utils/adt/datetime.c#L733-L1422) and [`DecodeTimeOnly`](https://github.com/postgres/postgres/blob/a995b371ae29de2d38c4b7881cf414b1560e9746/src/backend/utils/adt/datetime.c#L1701-L2325).

I don't expect that we'll need to be bug-for-bug compatible with the Postgres code anytime soon, but it's probably where we'll want to end up. That is, I don't think there is anything immediately actionable here, but we should study the Postgres implementation thoroughly before embarking on any refactors.

/cc @sploiselle @quodlibetor

**Comment by petrosagg** [comment]
I studied the implementation of date/time handling in Postgres and Materialize to try and tackle this issue.

In order to reach parity I see two major pieces of work. The first is being able to parse all the different weird date and time formats that postgres supports and the second is timezone handling.

## Date/Time formats ##

### Implementation ###

As explained in the issue description, postgres splits the input string into tokens using the [ParseDateTime function](https://github.com/postgres/postgres/blob/a995b371ae29de2d38c4b7881cf414b1560e9746/src/backend/utils/adt/datetime.c#L505-L730) which is then fed to the various `Decode*` functions.

The type of fields (month/year/timezone/etc) are either determined right at the tokenization phase, or given a generic type (e.g string) and are interpreted later by the `Decode*` functions. The logic is therefore spread across multiple functions. The interpretation is also parameterized by the `DateStyle` configuration variable that can be MDY, DMY, or YMD. This is used to disambiguate dates like `10/03/07` and `03/04/2020`.

The current materialize implementation follows a similar structure and even through there is a lot less pointer and buffer accounting if extended to support all the formats the logic would be similarly hard to follow.

Instead of writing this complex parser by hand I'd propose using [nom](https://github.com/Geal/nom) (or similar, it's just the one I have experience with) crate to write all the parsing rules. I'll try to write one that can parse the limited subset of dates that the current parser can just to compare code complexity and performance.

### Testing ###

In order to ensure a compatible implementation we can reuse the array of tests present in the postgresql source code. To aid this I have extracted all the positive testcases from the source code, inserted them in a table of the appropriate column type, and recorded the (input, parsed) pairs [in this gist](https://gist.github.com/petrosagg/590f79bde3438ed3ee233f96fea90bb1).

This can serve as a pretty good basis for ensuring the implementation can parse at least as much as postgres. I need to come back to this and extract negative testcases as well to make sure we're as strict as postgres.

### Steps ###

* [ ] Convert parser to nom and ensure no performance regression
* [ ] Parameterize parser by `DateStyle` and wire up rest of the project to use it
* [ ] `time` parsing parity
* [ ] `date` parsing parity
* [ ] `timestamp` parsing parity
* [ ] `timestamptz` parsing parity
* [ ] `time` parsing parity

## Timezones ##

Postgres supports three ways of specifying a timezone. The first is a simple numeric offset, the second one is using a timezone identifier as defined by IANA, and the third one is using a timezone abbreviation.

The good news is that postgres doesn't deviate at all from IANA and in fact vendors in a copy of the official tz database in each release (e.g [tzdata 2020d for postgres 13](https://github.com/postgres/postgres/commit/96ed2ae9360d9b89f695f00c2b6417c4e4d9fcba)). This means that in materialize we can use [chrono-tz](https://github.com/chronotope/chrono-tz) that does the same thing and integrates the database with chrono.

On the other hand, there is no global standard for timezone abbreviations and there are plenty of conflicts. The list of abbreviations are [parameterizeable in postgres](https://www.postgresql.org/docs/9.1/runtime-config-client.html#GUC-TIMEZONE-ABBREVIATIONS) with the [Default set here](https://github.com/postgres/postgres/blob/REL_13_1/src/timezone/tznames/Default). Support for abbreviations could be added upstream ([open issue](https://github.com/chronotope/chrono-tz/issues/44)) or just within materialize.

It's probably worth splitting this work into first integrating chrono-tz for the official timezones and then thinking about how to do the abbreviations.

### Steps ###

* [ ] Integrate chrono-tz for IANA timezone parsing
* [ ] Support abbreviated timezones

Does the above sound sensible to you? Is there is another avenue that I should take?

**Comment by quodlibetor** [comment]
Hi @petrosagg thanks for your interest in this!

You've done fantastic research and I think you're correct that this is definitely approachable, and most of the steps are steps we would love to take. In particular: supporting the full range of postgres input functions, getting more timezone support, and integrating chrono-tz.

We are pretty happy with our parser infrastructure, though, and if you're interested in doing some of this work we would prefer if it continues to use our hand-rolled parser. Hopefully knowing where we're going may allow us to approach the design with a coherent vision will allow us to tame some of the complexity.

The way that we have approached this kind of compatibility work is to create a full SQL Logic Test (SLT) file that has the full list of conversions, with the ones that are currently unsupported commented out. We could expand our current [`dates-times.slt`](https://github.com/MaterializeInc/materialize/blob/078c31efc7247ebef8c2befc1a155f603b0a979a/test/sqllogictest/dates-times.slt) with known-failing parses. I think that we would want to succeed on the full set of both positive and negative test cases before we were comfortable closing this issue.

I agree that timezone and fully-generic datetime parsing are orthogonal concerns, and could be tackled in any order. FWIW I maintain `chrono-tz` and would be happy to take a PR adding the ability to define a mapping from tz abbreviations to named timezones there.

If you're interested in taking on any part of this we can definitely continue! It would be exciting to make more progress on this issue!

**Comment by petrosagg** [comment]
@quodlibetor that sounds great, I think I talked a bit too soon with the parser. Proposing effectively a parser rewrite should come with some pretty compelling evidence for it, sorry about that. I'm happy to improve the existing parsing infrastructure with support for more datetime formats!

Thanks for the `dates-times.slt` pointer, I had only seen the tests under the repr crate.

>  We could expand our current dates-times.slt with known-failing parses. I think that we would want to succeed on the full set of both positive and negative test cases before we were comfortable closing this issue.

Sounds reasonable to me 

> If you're interested in taking on any part of this we can definitely continue! It would be exciting to make more progress on this issue!

I'll try to add a couple more datetime formats to get the ball rolling and take a look at the chrono-tz package

**Comment by quodlibetor** [comment]
> Proposing effectively a parser rewrite

I think that your attitude of "we know that this is going to become complex, let's try and use something that should help tame that complexity" makes lots of sense! We just happen to believe (from experience) that it will be hard to beat a hand-rolled parser for our long-term maintenance.

> I'll try to add a couple more datetime formats to get the ball rolling and take a look at the chrono-tz package.

That sounds great! I'll keep an eye out for any pings from you, and you can feel free to drop by our community slack (linked from the top of https://materialize.com) if you want some more real-time conversation.

**Comment by kthejoker** [comment]
Kind of a sidebar question: the default Materialize docker instance is configured with the DateStyle parameter to be read only.

This means you can't use the standard PostGres ODBC Driver for Materialize since its test query for all connections is "SET DateStyle = 'ISO';SET extra_float_digits = 2;"

Resulting in the error message 

"parameter DateStyle is read only"

Which, amazingly, shows up in precisely **0** Google hits - until now! So I don't even know how to modify the Materialize Docker image code or configuration in any way to make the parameter writeable.

Any help or thoughts?

**Comment by benesch** [comment]
Are you sure? It should be in v0.5.3. Have you run `docker pull` lately?

```
docker pull materialize/materialized:latest
```

**Comment by kthejoker** [comment]
I did do a pull, but I must've messed something up with restarting the container, as I did a full down and up and I see it's working as expected now, thanks, @benesch
